File: src/schema.ts
import { z } from "zod";

export const JobStatusSchema = z.enum([
  "working",
  "needs_input",
  "idle",
  "completed",
  "failed",
  "stopped",
]);

export const ProcessStateSchema = z.enum(["alive", "exited", "sleeping", "unknown"]);

export const PrRefSchema = z.object({
  url: z.string(),
  owner: z.string(),
  repo: z.string(),
  number: z.number(),
  status: z.string().default("unknown"),
});

export const BlockingRequestSchema = z.object({
  type: z.string(),
  message: z.string(),
  event: z.unknown().optional(),
  createdAt: z.string(),
});

export const JobSchema = z.object({
  id: z.string(),
  provider: z.literal("codex"),
  codexThreadId: z.string().nullable(),
  title: z.string(),
  initialPrompt: z.string(),
  repoRoot: z.string(),
  cwd: z.string(),
  dispatchCwd: z.string(),
  worktreePath: z.string().nullable(),
  worktreeBranch: z.string().nullable(),
  model: z.string().nullable(),
  profile: z.string().nullable(),
  approvalPolicy: z.enum(["untrusted", "on-request", "never"]),
  sandbox: z.enum(["read-only", "workspace-write", "danger-full-access"]),
  status: JobStatusSchema,
  processState: ProcessStateSchema,
  pid: z.number().nullable(),
  activeWorkerPid: z.number().nullable(),
  pinned: z.boolean(),
  manualOrder: z.number().nullable(),
  archived: z.boolean(),
  deleted: z.boolean(),
  lastSummary: z.string().nullable(),
  lastOutput: z.string().nullable(),
  blockingRequest: BlockingRequestSchema.nullable(),
  prRefs: z.array(PrRefSchema),
  createdAt: z.string(),
  updatedAt: z.string(),
  completedAt: z.string().nullable(),
  exitCode: z.number().nullable(),
  error: z.string().nullable().optional(),
});

export const StoreSchema = z.object({
  version: z.number(),
  jobs: z.record(z.string(), JobSchema),
  preferences: z.record(z.string(), z.unknown()),
});

export const InboxMessageSchema = z.object({
  type: z.literal("reply"),
  prompt: z.string(),
  timestamp: z.string().optional(),
});

export type JobStatus = z.infer<typeof JobStatusSchema>;
export type ProcessState = z.infer<typeof ProcessStateSchema>;
export type PrRef = z.infer<typeof PrRefSchema>;
export type BlockingRequest = z.infer<typeof BlockingRequestSchema>;
export type Job = z.infer<typeof JobSchema>;
export type Store = z.infer<typeof StoreSchema>;
export type InboxMessage = z.infer<typeof InboxMessageSchema>;

export type JobPatch = Partial<Job>;

File: crates/agentview-core/src/schema.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JobStatus {
    Working,
    NeedsInput,
    Idle,
    Completed,
    Failed,
    Stopped,
}

impl JobStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Working => "working",
            Self::NeedsInput => "needs_input",
            Self::Idle => "idle",
            Self::Completed => "completed",
            Self::Failed => "failed",
            Self::Stopped => "stopped",
        }
    }
}

impl std::fmt::Display for JobStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProcessState {
    Alive,
    Exited,
    Sleeping,
    Unknown,
}

impl ProcessState {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Alive => "alive",
            Self::Exited => "exited",
            Self::Sleeping => "sleeping",
            Self::Unknown => "unknown",
        }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum JobBackend {
    #[default]
    FallbackExec,
    AppServer,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PrRef {
    pub url: String,
    pub owner: String,
    pub repo: String,
    pub number: u64,
    pub status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BlockingRequest {
    #[serde(rename = "type")]
    pub kind: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub event: Option<Value>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Job {
    pub id: String,
    pub provider: String,
    #[serde(default)]
    pub backend: JobBackend,
    #[serde(rename = "codexThreadId")]
    pub codex_thread_id: Option<String>,
    #[serde(
        rename = "codexTurnId",
        default,
        skip_serializing_if = "Option::is_none"
    )]
    pub codex_turn_id: Option<String>,
    pub title: String,
    #[serde(rename = "initialPrompt")]
    pub initial_prompt: String,
    #[serde(rename = "repoRoot")]
    pub repo_root: String,
    pub cwd: String,
    #[serde(rename = "dispatchCwd")]
    pub dispatch_cwd: String,
    #[serde(rename = "worktreePath")]
    pub worktree_path: Option<String>,
    #[serde(rename = "worktreeBranch")]
    pub worktree_branch: Option<String>,
    pub model: Option<String>,
    pub profile: Option<String>,
    #[serde(rename = "approvalPolicy")]
    pub approval_policy: String,
    pub sandbox: String,
    pub status: JobStatus,
    #[serde(rename = "processState")]
    pub process_state: ProcessState,
    pub pid: Option<u32>,
    #[serde(rename = "activeWorkerPid")]
    pub active_worker_pid: Option<u32>,
    pub pinned: bool,
    #[serde(rename = "manualOrder")]
    pub manual_order: Option<i64>,
    pub archived: bool,
    pub deleted: bool,
    #[serde(rename = "lastSummary")]
    pub last_summary: Option<String>,
    #[serde(rename = "lastOutput")]
    pub last_output: Option<String>,
    #[serde(rename = "blockingRequest")]
    pub blocking_request: Option<BlockingRequest>,
    #[serde(rename = "prRefs")]
    pub pr_refs: Vec<PrRef>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(rename = "updatedAt")]
    pub updated_at: String,
    #[serde(rename = "completedAt")]
    pub completed_at: Option<String>,
    #[serde(rename = "exitCode")]
    pub exit_code: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Store {
    pub version: u32,
    pub jobs: BTreeMap<String, Job>,
    #[serde(default)]
    pub preferences: BTreeMap<String, Value>,
}

impl Default for Store {
    fn default() -> Self {
        Self {
            version: 1,
            jobs: BTreeMap::new(),
            preferences: BTreeMap::new(),
        }
    }
}
